File: 22.linq-workshop-kata/java/src/test/java/linq_workshop/realworld/RealWorldTest.java
package linq_workshop.realworld;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import linq_workshop.realworld.watch.Component;
import linq_workshop.realworld.watch.ComponentType;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.InputStream;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;

class RealWorldTest {

    private static Component watch;

    @BeforeAll
    static void globalSetup() throws IOException {
        watch = deserializeWatch();
    }

    private static Component deserializeWatch() throws IOException {
        var mapper = new ObjectMapper()
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

        try (InputStream input = RealWorldTest.class.getClassLoader().getResourceAsStream("watch.json")) {
            return mapper.readValue(input, Component.class);
        }
    }

    @Test
    void allComponents() {
        int totalCount = 0;

        assertThat(totalCount).isEqualTo(564);
    }

    @Test
    void markers() {
        List<Component> markers = null;

        assertThat(markers).hasSize(14);
        assertThat(markers.stream().map(c -> c.name).toList()).containsExactlyInAnyOrder(
            "Markers 305",
            "Markers 142",
            "Markers 243",
            "Markers 269",
            "Markers 288",
            "Markers 333",
            "Markers 338",
            "Markers 345",
            "Markers 426",
            "Markers 457",
            "Markers 468",
            "Markers 472",
            "Markers 494",
            "Markers 540"
        );
    }

    @Test
    void findAllComponentStartingWith() {
        long countDial = 0;

        assertThat(countDial).isEqualTo(18);
    }

    @Test
    void countComponentByTypes() {
        Map<ComponentType, Long> types = null;

        assertThat(types.get(ComponentType.Oscillator)).isEqualTo(25);
        assertThat(types.get(ComponentType.BalanceWheel)).isEqualTo(18);
        assertThat(types.get(ComponentType.Wheel)).isEqualTo(19);
    }

    @Test
    void mostComplex() {
        Component mostComplexComponent = null;
        int componentsInsideBezel = 0;

        assertThat(mostComplexComponent.name).isEqualTo("Bezel 1");
        assertThat(componentsInsideBezel).isEqualTo(38);
    }

    @Test
    void averageComponentsByType() {
        Map<ComponentType, Double> averageSubcomponentsByType = null;
    }
}

File: 22.linq-workshop-kata/ts/tests/03-real-world/real-world.test.ts
import { beforeAll, describe, expect, it } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";
import { Component, RawComponent, parseComponent } from "../../src/03-real-world/watch/component";
import { ComponentType } from "../../src/03-real-world/watch/component-type";
import { sameMembers } from "../support/collections";

const currentDir = dirname(fileURLToPath(import.meta.url));

describe("RealWorld", () => {
  let watch: Component;

  beforeAll(() => {
    const raw = JSON.parse(readFileSync(join(currentDir, "watch.json"), "utf-8")) as RawComponent;
    watch = parseComponent(raw);
  });

  it("allComponents", () => {
    const totalCount = 0;

    expect(totalCount).toBe(564);
  });

  it("markers", () => {
    const markers: Component[] | undefined = undefined;

    expect(markers!).toHaveLength(14);
    sameMembers(
      markers!.map((m) => m.name),
      [
        "Markers 305",
        "Markers 142",
        "Markers 243",
        "Markers 269",
        "Markers 288",
        "Markers 333",
        "Markers 338",
        "Markers 345",
        "Markers 426",
        "Markers 457",
        "Markers 468",
        "Markers 472",
        "Markers 494",
        "Markers 540",
      ]
    );
  });

  it("findAllComponentStartingWith", () => {
    const countDial = 0;

    expect(countDial).toBe(18);
  });

  it("countComponentByTypes", () => {
    const types: Map<ComponentType, number> | undefined = undefined;

    expect(types!.get(ComponentType.Oscillator)).toBe(25);
    expect(types!.get(ComponentType.BalanceWheel)).toBe(18);
    expect(types!.get(ComponentType.Wheel)).toBe(19);
  });

  it("mostComplex", () => {
    const mostComplexComponent: Component | undefined = undefined;
    const componentsInsideBezel = 0;

    expect(mostComplexComponent!.name).toBe("Bezel 1");
    expect(componentsInsideBezel).toBe(38);
  });

  it("averageComponentsByType", () => {
    const averageSubcomponentsByType: Map<ComponentType, number> | undefined = undefined;
  });
});
